src/olympia/reviewers/views.py

```diff
@@ -1240,43 +1241,85 @@ def flags(self, request, **kwargs):
         # Permission classes disallow access to non-public/unlisted add-ons
         # unless logged in as a reviewer/addon owner/admin, so we don't have to
         # filter the base queryset here.
-        return Version.unfiltered.all()
+        return (
+            Version.unfiltered
+            .filter(addon=self.get_addon_object())
+            .order_by('-created'))
 
-    def get_serializer_context(self):
-        context = super(BrowseVersionViewSet, self).get_serializer_context()
-        context.update({
-            'file': self.request.GET.get('file', None)
-        })
-        return context
+    def get_addon_object(self):
+        return get_object_or_404(Addon, pk=self.kwargs.get('addon_pk'))
 
-    def check_object_permissions(self, request, obj):
-        """
-        Check if the request should be permitted for a given version.
+    def get_object(self):
+        qset = self.filter_queryset(self.get_queryset())
+
+        filter_kwargs = {self.lookup_field: self.kwargs[self.lookup_url_kwarg]}
+
+        obj = get_object_or_404(qset, **filter_kwargs)
 
-        This calls the DRF implementation but forwards the respective Add-on
-        as `obj` since most of our permission classes are based on an Add-on
-        being passed around.
-        """
         # If the instance is marked as deleted and the client is not allowed to
         # see deleted instances, we want to return a 404, behaving as if it
         # does not exist.
         if obj.deleted and not (
                 GroupPermission(amo.permissions.ADDONS_VIEW_DELETED).
-                has_object_permission(request, self, obj.addon)):
+                has_object_permission(self.request, self, obj.addon)):
             raise http.Http404
 
-        return (
-            super(BrowseVersionViewSet, self)
-            .check_object_permissions(request, obj.addon))
+        # Now we can checking permissions
+        self.check_object_permissions(self.request, obj)
+
+        return obj
+
+    def check_permissions(self, request):
+        if self.action == u'list':
+            # When listing DRF doesn't explicitly check for object permissions
+            # but here we need to do that against the parent add-on.
+            # So we're calling check_object_permission() ourselves,
+            # which will pass down the addon object directly.
+            return (
+                super(ReviewAddonVersionViewSet, self)
+                .check_object_permissions(request, self.get_addon_object()))
+
+        super(ReviewAddonVersionViewSet, self).check_permissions(request)
+
+    def check_object_permissions(self, request, obj):
+        """Check permissions against the parent add-on object."""
+        return super(ReviewAddonVersionViewSet, self).check_object_permissions(
+            request, obj.addon)
+
+    def list(self, request, *args, **kwargs):
+        """Return all (re)viewable versions for this add-on.
+
+        Full list, no pagination."""
```

Comment:
Why no pagination? This seems like a bad idea, some add-ons have a lot of versions...

Reply:
Because that list is being shown in a drop-down of some kind and I personally don't think fetching the versions through pagination is a worthwhile undertaking.

I don't know how the current file-viewer handles huge amounts of versions but I'd guess, it just looks very stupid.

Reply:
yeah... (see https://reviewers.addons.mozilla.org/en-US/firefox/files/compare/1683453...1679753/ for 83 versions - it's slow)

But if it _only_ every returns minimal information (like the serializer does now) then it's probably okay.  It'll just get terrible quickly if we start returning nested serializers and other objects (e.g. translations).  Maybe add a comment to the serializer for future us to not add anything crazy to it.

Reply:
I could even make the query a bit quicker by reducing the number of fields we are selecting to the ones we actually need which should be fine then.

Reply:
`id`, `channel` `version` shouldn't be terrible.

Reply:
Best way to make it super fast (besides caching and accepting that the first request is going to be slow) would be to only select columns that are part of the same composite index ; that way MySQL wouldn't need to fetch the data from disk. Unfortunately we don't have an index that match those fields...

Reply:
I personally don't think we have to make it super-fast, fast enough should be fine as long as the loading doesn't time-out and stays responsive for the user.